### ISSUE TITLE
Typescript AuthToken is missing signInDetails

### ISSUE THREAD

**Issue by ChristopherGabba** [aws-amplify/amplify-js#12868]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React Native

### Amplify APIs

Authentication

### Amplify Version

v6

### Amplify Categories

auth

### Backend

Amplify CLI

### Environment information

<details>

```
# Put output below this line

  System:
    OS: macOS 14.2.1
    CPU: (10) arm64 Apple M2 Pro
    Memory: 321.95 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.7.0 - /opt/homebrew/bin/node
    npm: 10.1.0 - /opt/homebrew/bin/npm
    Watchman: 2023.09.04.00 - /opt/homebrew/bin/watchman
  Browsers:
    Safari: 17.2.1
  npmPackages:
    @aws-amplify/react-native: ^1.0.12 => 1.0.12 
    @babel/core: ^7.20.0 => 7.23.7 
    @babel/plugin-proposal-export-namespace-from: ^7.18.9 => 7.18.9 
    @babel/preset-env: ^7.20.0 => 7.23.8 
    @babel/runtime: ^7.20.0 => 7.23.8 
    @expo-google-fonts/m-plus-1p: ^0.2.3 => 0.2.3 
    @expo-google-fonts/montserrat: ^0.2.3 => 0.2.3 
    @expo-google-fonts/space-grotesk: ^0.2.2 => 0.2.3 
    @expo/config-plugins: ~7.2.2 => 7.2.5 (5.0.4)
    @expo/metro-config: ^0.10.7 => 0.10.7 
    @gorhom/bottom-sheet: ^4.4.7 => 4.6.0 
    @likashefqet/react-native-image-zoom: ^2.1.1 => 2.1.1 
    @react-native-async-storage/async-storage: ^1.18.2 => 1.21.0 
    @react-native-clipboard/clipboard: ^1.11.2 => 1.13.2 
    @react-native-community/cli-platform-ios: ^8.0.2 => 8.0.6 (11.3.7)
    @react-native-community/netinfo: ^9.3.10 => 9.5.0 
    @react-navigation/bottom-tabs: ^6.3.2 => 6.5.11 
    @react-navigation/native: ^6.0.2 => 6.1.9 
    @react-navigation/native-stack: ^6.0.2 => 6.9.17 
    @rnx-kit/metro-config: ^1.3.5 => 1.3.14 
    @rnx-kit/metro-resolver-symlinks: ^0.1.26 => 0.1.34 
    @sentry/react-native: 5.10.0 => 5.10.0 
    @shopify/flash-list: 1.4.3 => 1.4.3 
    @types/i18n-js: 3.8.2 => 3.8.2 
    @types/jest: ^29.2.1 => 29.5.11 
    @types/lodash.filter: ^4.6.9 => 4.6.9 
    @types/react: ~18.2.14 => 18.2.48 
    @types/react-test-renderer: ^18.0.0 => 18.0.7 
    @typescript-eslint/eslint-plugin: ^5.59.0 => 5.62.0 
    @typescript-eslint/parser: ^5.59.0 => 5.62.0 
    HelloWorld:  0.0.1 
    apisauce: 2.1.5 => 2.1.5 
    aws-amplify: ^6.0.12 => 6.0.12 
    aws-amplify/adapter-core:  undefined ()
    aws-amplify/analytics:  undefined ()
    aws-amplify/analytics/kinesis:  undefined ()
    aws-amplify/analytics/kinesis-firehose:  undefined ()
    aws-amplify/analytics/personalize:  undefined ()
    aws-amplify/analytics/pinpoint:  undefined ()
    aws-amplify/api:  undefined ()
    aws-amplify/api/server:  undefined ()
    aws-amplify/auth:  undefined ()
    aws-amplify/auth/cognito:  undefined ()
    aws-amplify/auth/cognito/server:  undefined ()
    aws-amplify/auth/enable-oauth-listener:  undefined ()
    aws-amplify/auth/server:  undefined ()
    aws-amplify/datastore:  undefined ()
    aws-amplify/in-app-messaging:  undefined ()
    aws-amplify/in-app-messaging/pinpoint:  undefined ()
    aws-amplify/push-notifications:  undefined ()
    aws-amplify/push-notifications/pinpoint:  undefined ()
    aws-amplify/storage:  undefined ()
    aws-amplify/storage/s3:  undefined ()
    aws-amplify/storage/s3/server:  undefined ()
    aws-amplify/storage/server:  undefined ()
    aws-amplify/utils:  undefined ()
    axios: ^1.5.0 => 1.6.5 (0.21.4)
    babel-jest: ^29.2.1 => 29.7.0 
    babel-loader: 8.2.5 => 8.2.5 
    babel-plugin-root-import: ^6.6.0 => 6.6.0 
    cheerio: ^1.0.0-rc.12 => 1.0.0-rc.12 
    core-util-is:  1.0.1 
    date-fns: ^2.29.2 => 2.30.0 
    eslint: 8.17.0 => 8.17.0 
    eslint-config-prettier: 8.5.0 => 8.5.0 
    eslint-config-standard: 17.0.0 => 17.0.0 
    eslint-plugin-import: 2.26.0 => 2.26.0 
    eslint-plugin-n: ^15.0.0 => 15.7.0 
    eslint-plugin-node: 11.1.0 => 11.1.0 
    eslint-plugin-promise: 6.0.0 => 6.0.0 
    eslint-plugin-react: 7.30.0 => 7.30.0 
    eslint-plugin-react-native: 4.0.0 => 4.0.0 
    expo: ^49.0.21 => 49.0.21 
    expo-application: ~5.3.0 => 5.3.1 
    expo-blur: ~12.4.1 => 12.4.1 
    expo-build-properties: ~0.8.3 => 0.8.3 
    expo-clipboard: ~4.3.1 => 4.3.1 
    expo-config-plugin-ios-share-extension: ^0.0.4 => 0.0.4 
    expo-constants: ~14.4.2 => 14.4.2 
    expo-contacts: ~12.2.0 => 12.2.0 
    expo-dev-client: ~2.4.11 => 2.4.12 
    expo-device: ~5.4.0 => 5.4.0 
    expo-file-system: ~15.4.4 => 15.4.5 
    expo-font: ~11.4.0 => 11.4.0 
    expo-image: ~1.3.5 => 1.3.5 
    expo-image-picker: ~14.3.2 => 14.3.2 
    expo-linear-gradient: ~12.3.0 => 12.3.0 
    expo-linking: ~5.0.2 => 5.0.2 
    expo-localization: ~14.3.0 => 14.3.0 
    expo-media-library: ~15.4.1 => 15.4.1 
    expo-notifications: ~0.20.1 => 0.20.1 
    expo-secure-store: ~12.3.1 => 12.3.1 
    expo-splash-screen: ^0.20.5 => 0.20.5 
    expo-status-bar: ~1.6.0 => 1.6.0 
    expo-store-review: ~6.4.0 => 6.4.0 
    expo-video-thumbnails: ~7.4.0 => 7.4.0 
    fbjs-scripts: 3.0.1 => 3.0.1 
    i18n-js: 3.9.2 => 3.9.2 
    inherits:  2.0.1 
    isarray:  0.0.1 
    jest: ^29.2.1 => 29.7.0 
    jest-circus: 29 => 29.7.0 
    jest-environment-node: 29 => 29.7.0 
    jest-expo: ^49.0.0 => 49.0.0 
    jsdom: ^22.1.0 => 22.1.0 (20.0.3)
    jsdom-jscore-rn: ^0.1.8 => 0.1.8 
    lodash: ^4.17.21 => 4.17.21 
    lodash.filter: ^4.6.0 => 4.6.0 
    lottie-react-native: ^5.1.6 => 5.1.6 
    metro-config: 0.76.8 => 0.76.8 
    metro-source-map: 0.75.1 => 0.75.1 (0.76.8)
    mobx: 6.6.0 => 6.6.0 
    mobx-react-lite: 3.4.0 => 3.4.0 
    mobx-state-tree: 5.1.5 => 5.1.5 
    mocha: ^10.2.0 => 10.2.0 
    nwmatcher:  1.4.3 
    patch-package: ^6.4.7 => 6.5.1 
    path-browserify:  0.0.0 
    postinstall-prepare: 1.0.1 => 1.0.1 
    prettier: 2.8.8 => 2.8.8 (3.2.4)
    query-string: ^7.0.1 => 7.1.3 (6.10.1)
    querystring:  0.2.0 
    react: 18.2.0 => 18.2.0 
    react-devtools-core: 4.24.7 => 4.24.7 (4.28.5)
    react-dom: 18.2.0 => 18.2.0 
    react-native: 0.72.6 => 0.72.6 
    react-native-bootsplash: ^5.0.2 => 5.2.2 
    react-native-compressor: ^1.8.23 => 1.8.23 
    react-native-device-info: ^10.12.0 => 10.12.0 
    react-native-dots-pagination: ^0.3.1 => 0.3.1 
    react-native-element-dropdown: ^2.9.0 => 2.10.1 
    react-native-fs: ^2.20.0 => 2.20.0 
    react-native-gesture-handler: ~2.12.0 => 2.12.1 
    react-native-get-random-values: ~1.9.0 => 1.9.0 
    react-native-mmkv: ^2.11.0 => 2.11.0 
    react-native-pager-view: 6.2.0 => 6.2.0 
    react-native-reanimated: ~3.3.0 => 3.3.0 
    react-native-receive-sharing-intent: ^2.0.0 => 2.0.0 
    react-native-render-html: ^6.3.4 => 6.3.4 
    react-native-safe-area-context: 4.6.3 => 4.6.3 
    react-native-screens: ~3.22.0 => 3.22.1 
    react-native-share-menu: ^6.0.0 => 6.0.0 
    react-native-shimmer-placeholder: ^2.0.9 => 2.0.9 
    react-native-static-safe-area-insets: ^2.2.0 => 2.2.0 
    react-native-touchable-scale: ^2.2.0 => 2.2.0 
    react-native-url-polyfill: ^2.0.0 => 2.0.0 
    react-native-video: ^6.0.0-beta.4 => 6.0.0-beta.4 
    react-native-vision-camera: ^3.6.6 => 3.8.0 
    react-native-volume-manager: ^1.10.0 => 1.10.0 
    react-native-web: ~0.19.6 => 0.19.10 
    react-native-webview: 13.2.2 => 13.2.2 
    react-native-youtube-iframe: ^2.3.0 => 2.3.0 
    react-test-renderer: 18.2.0 => 18.2.0 
    reactotron-core-client: ^2.8.10 => 2.8.11 (2.8.10)
    reactotron-mst: 3.1.4 => 3.1.4 
    reactotron-react-js: ^3.3.7 => 3.3.9 
    reactotron-react-native: 5.0.3 => 5.0.3 
    regenerator-runtime: ^0.13.4 => 0.13.11 (0.14.1)
    sentry-expo: ~7.1.0 => 7.1.1 
    string_decoder:  0.10.31 
    ts-jest: 29 => 29.1.1 
    typescript: ^4.9.4 => 5.3.3 
    urlmaster:  0.2.15 
  npmGlobalPackages:
    @aws-amplify/cli-internal: 12.10.0
    @aws-amplify/cli: 12.10.1
    @react-native-community/netinfo: 9.4.1
    eas-cli: 7.0.0
    expo-cli: 6.3.10
    firebase-tools: 11.24.1
    n: 9.1.0
    node: 20.6.0
    npm: 10.3.0
    pod-install: 0.1.39
    react-native-spinkit: 1.5.1

```

</details>

### Describe the bug

Simple bug, getting typescript error of type AuthToken:
`Property 'signInDetails' does not exist on type 'AuthTokens'`
Yet when I console log, I get this:
```json
{
"credentials": {
      "accessKeyId": "XXXXXXXXX", 
      "expiration": "XXXXXXXXXXXXX",
      "secretAccessKey": "XXXXXXX", 
      "sessionToken": "XXXXXXXXX"
 }, 
"identityId": "us-east-XXXXX", 
"tokens": {
     "accessToken": {
           "payload":" XXXXXX", 
           "toString": "[Function toString]",
      }, 
      "idToken": {
             "payload": "XXXX"
             "toString": "XXXXXX"
       }, 
       "signInDetails": {    // < --------- This branch is missing from the Typescript aka not typesafe
              "authFlowType": "USER_SRP_AUTH", 
              "loginId": "Thegabbanator"
       }
 }, 
"userSub": "63d73ce8-526e-4b4d-8a92-15df5409e371"
}

```

### Expected behavior

Simple enough above.... 

### Reproduction steps

Using custom hook that I made: 
Use: 
```typescript
const session = await fetchAuthSession()
session.tokens?.signInDetÏails?.loginId       // <------- Getting typescript error here
```

### Code Snippet

```typescript
 useEffect(() => {
    ;(async function manageAuth() {
      try {
        const session = await fetchAuthSession()
        const userAttributes = await fetchUserAttributes()

        setCurrentUser({
          id: session.userSub as string,
          username: session.tokens?.signInDetÏails?.loginId,  // <------- Getting typescript error here
          firstName: userAttributes.given_name,
          lastName: userAttributes.family_name,
          profileImage: userAttributes.picture,
          birthdate: userAttributes.birthdate,
          phoneNumber: userAttributes.phone_number,
        })

      } catch (e) {
        reportCrash(e)
      }
    })()
  }, [])
```

### Log output

<details>

```
// Put your logs below this line


```

</details>

### Mobile Device

iPhone 12

**Comment by nadetastic** [comment]
Hi @ChristopherGabba thanks for opening this issue. I'm able to reproduce this behavior. Let me discuss this with the team and will follow up soon. In the meantime, let me know if you have any questions.

**Comment by adeelejaz** [comment]
+1

I’m seeing similar issue in non-typescript Next.js project. signedInDetails is missing on the response from currentUser only for some users.

**Comment by ChristopherGabba** [comment]
@nadetastic, there is another Typescript bug that is very similar. 
```typescript
const userAttributes = await fetchUserAttributes()
```
This function returns the attributes of the authenticated user correctly, but every single attribute is optional `?`. Even the sub attribute which is definitely mandatory.

This leads to more typescript warnings in my project. In reality, it should return the authenticated user's attributes that are mandatory as non-optional. 

For instance, here is my user pool property and the grayed / checked attributes I've made mandatory, yet fetchUserAttributes is showing them as optional for me.
![image](https://github.com/aws-amplify/amplify-js/assets/122328438/e5943158-c301-410d-ad63-f4bd2f353ddc)

I know ```fetchUserAttributes``` in v6 is different from the old ```Auth.currentAuthenticatedUser()```, but at least the v5 function had the typing optionality correct.

**Comment by israx** [comment]
hey @ChristopherGabba `fetchUserAttributes` will return all attributes as optional, the reason is that you can configure Cognito `UserPool` attributes as required and non-required, so at build time there is not a clean way to infer the user's configuration, one option is to make `fetchUserAttributes`, `signUp` and other APIs generic and enable the developer to pass a generic type value and enforce required attributes. Do you mind opening a feature request with us ?

**Comment by ChristopherGabba** [comment]
@israx done: https://github.com/aws-amplify/amplify-js/issues/12901
Thanks!

**Comment by jimblanc** [comment]
@ChristopherGabba The fix for this went out with Amplify v6.0.18. Please let us know if you run into any issues!
